wrap/grpc.go

```diff
@@ -0,0 +1,176 @@
+type ProtoService struct {
+	Name    string
+	Methods []ServiceMethod
+}
+
+// WrapperData is the template data structure.
+type WrapperData struct {
+	Package  string
+	Service  string
+	Methods  []ServiceMethod
+	Requests []string
+}
+
+func GenerateWrapper(ctx *gofr.Context) (any, error) {
+	protoPath := ctx.Param("proto")
+	if protoPath == "" {
+		return nil, ErrNoProtoFile
+	}
+
+	file, err := os.Open(protoPath)
+	if err != nil {
+		ctx.Errorf("Failed to open proto file: %v", err)
+
+		return nil, ErrOpeningProtoFile
+	}
+
+	defer file.Close()
+
+	parser := proto.NewParser(file)
+
+	definition, err := parser.Parse()
+	if err != nil {
+		ctx.Errorf("Failed to parse proto file: %v", err)
+
+		return nil, ErrFailedToParseProto
+	}
+
+	var (
+		// Extracting package and project path from go_package option.
+		packageName, projectPath = getPackageAndProject(definition)
+		services                 []ProtoService
+	)
+
+	// Extract the services.
+	getServices(definition, services)
+
+	for _, service := range services {
+		wrapperData := WrapperData{
+			Package:  packageName,
+			Service:  service.Name,
+			Methods:  service.Methods,
+			Requests: uniqueRequestTypes(service.Methods),
+		}
+
+		generatedCode := generateWrapperCode(ctx, &wrapperData)
+		if generatedCode == "" {
+			return nil, ErrGeneratingWrapper
+		}
+
+		outputFilePath := fmt.Sprintf("%s/%s.gofr.go", projectPath, strings.ToLower(service.Name))
+
+		err := os.WriteFile(outputFilePath, []byte(generatedCode), filePerm)
+		if err != nil {
+			ctx.Errorf("Failed to write file %s: %v", outputFilePath, err)
+
+			return nil, ErrWritingWrapperFile
+		}
+
+		fmt.Printf("Generated wrapper for service %s at %s\n", service.Name, outputFilePath)
+	}
+
+	return "Successfully generated all wrappers for gRPC services", nil
+}
+
+// Extract unique request types from methods.
+func uniqueRequestTypes(methods []ServiceMethod) []string {
+	requests := make(map[string]bool)
+	req := make([]string, 0)
+
+	for _, method := range methods {
+		if !method.Streaming {
+			requests[method.Request] = true
+		}
+	}
+
+	for method := range requests {
+		req = append(req, method)
+	}
+
+	return req
+}
+
+// Generate wrapper code using the template.
+func generateWrapperCode(ctx *gofr.Context, data *WrapperData) string {
+	var buf bytes.Buffer
+
+	tmplInstance := template.Must(template.New("wrapper").Parse(tmpl))
+
+	err := tmplInstance.Execute(&buf, data)
+	if err != nil {
+		ctx.Errorf("Template execution failed: %v", err)
+
+		return ""
+	}
+
+	return buf.String()
+}
+
+func getPackageAndProject(definition *proto.Proto) (projectPath, packageName string) {
+	proto.Walk(definition,
+		proto.WithOption(func(opt *proto.Option) {
+			if opt.Name == "go_package" {
+				projectPath = opt.Constant.Source[:len(opt.Constant.Source)-1]
+				packageName = path.Base(opt.Constant.Source)
+			}
+		}),
+	)
+
+	return projectPath, packageName
+}
+
+func getServices(definition *proto.Proto, services []ProtoService) {
```

Comment:
Instead of passing the services as a slice in this `getServices` function, isn't it better to return the **[]ProtoService**? We can have a local slice services and append to it and return at end.